src/combinatorics/ssyt.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2022 Jake Levinson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jake Levinson
+-/
+import combinatorics.young_diagram
+
+/-!
+# Semistandard Young tableaux
+
+A semistandard Young tableau is a filling of a Young diagram by natural numbers, such that
+the entries are weakly increasing left-to-right along rows (i.e. for fixed `i`), and
+strictly-increasing top-to-bottom along columns (i.e. for fixed `j`).
+
+An example of an SSYT of shape `μ = [4, 2, 1]` is:
+
+```text
+0 0 0 2
+1 1
+2
+```
+
+We represent an SSYT as a function `ℕ → ℕ → ℕ`, which is required to be zero for all pairs
+`(i, j) ∉ μ` and to satisfy the row-weak and column-strict conditions on `μ`.
+
+
+## Main definitions
+
+- `ssyt (μ : young_diagram)` : semistandard Young tableaux of shape μ
+- `ssyt.highest_weight (μ : young_diagram)`: the semistandard Young tableau whose `i`th row
+  consists entirely of `i`s, for each `i`.
+
+## Tags
+
+Semistandard Young tableau
+
+## References
+
+<https://en.wikipedia.org/wiki/Young_tableau>
+
+-/
+
+/-- A semistandard Young tableau (SSYT) is a filling of the cells of a Young diagram by natural
+numbers, such that the entries in each row are weakly increasing (left to right), and the entries
+in each column are strictly increasing (top to bottom).
+
+Here, an SSYT is represented as an unrestricted function `ℕ → ℕ → ℕ`, but which is required to
+vanish outside `μ`. --/
+structure ssyt (μ : young_diagram) :=
```

Comment:
I think you should change `row_weak` to `row_weak'` and `col_strict` to `col_strict'` to hint at the fact that these will be restated in terms of the `has_coe_to_fun` instance.

Reply:
`row_weak` and `col_strict` don't actually get restated anywhere, they're just used as is. Are you saying I should add versions of them using that are restated in terms of the `has_coe_to_fun` instance?

Reply:
Yes, I'm meaning to say that we should have a restatement in terms of the `has_coe_to_fun` instances. The cost to using `has_coe_to_fun` is that you need to ensure the "public-facing API" is all in terms of this coercion. In other words, it has to be as if you applied the `to_fun_eq_coe` simp lemma to everything.

How I've see this handled is that the `structure` uses primes in their names and then the lemmas are restated without the primes.

(What I was thinking about when I wrote it is that you have `row_weak_of_le` and `col_weak`, which are sort of restatements.)